[R-package] added R linting and changed R code to comma-first (fixes #2373)

jameslamb opened the pull request:
In this PR, I propose adding a linter for code in the R package. For now I propose just tacking this on to Travis until I figure out #2353 (I have a support case pending with GitHub support for what I think is a bug in GitHub Actions).

See more information behind this PR in #2373 . 

This block in `.ci/lint_r_code.R` documents the key linters I've chosen to use from the `lintr` package.

```r
LINTERS_TO_USE <- list(
    "open_curly" = lintr::open_curly_linter
    , "closed_curly" = lintr::closed_curly_linter
    , "tabs" = lintr::no_tab_linter
    , "spaces" = lintr::infix_spaces_linter
    , "trailing_blank" = lintr::trailing_blank_lines_linter
    , "trailing_white" = lintr::trailing_whitespace_linter
    , "long_lines" = lintr::line_length_linter(length = 120)
)
```

They express the following:

* `{` or `}` should never be on a line by themselves
* only spaces should be used for indentation (not tabs)
* infix operators should have spaces around them, so e.g. `rnorm(n = 5)` is preferred to `rnorm(n=5)`
* files should end in exactly one empty line
* files should never have trailing horizontal whitespace
* no files should have lines longer than 120 characters

These linters are cheap to run, as they don't require installation of the package or evaluation of any of the package's code. There is no clean "always use comma-first" liner that I'm aware of for R code, so that will have to be enforced by convention.

In the way I've set this up, this will be applied to all files with extension `.R` or `.r`.

The files introduced into `.ci/` are adopted from a set of scripts I maintain for re-use by various open source projects I work on (https://github.com/jameslamb/ci_tools/tree/master/r).

jameslamb commented:
`Microsoft.LightGBM` and `appveyor` are failing with this error that I'm pretty sure is unrelated to this PR:

```
ImportError: No module named 'win32api'
```

StrikerRUS commented on .ci/test.sh:
```diff
@@ -59,6 +59,14 @@ if [[ $TASK == "lint" ]]; then
+    conda install -c r \
+        r-argparse \
+        r-lintr
+    ./.ci/lint_r.sh $(pwd)
```
Use absolute path: `$BUILD_DIRECTORY/.ci/lint_r.sh`

StrikerRUS commented on .ci/lint_r.sh:
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Do we really need two script files just to perform the R lint check? I guess we don't need to create a separate `.sh` file and litter the `.ci` folder to just launch another one script.

jameslamb replied:
oh sure! I thought it was a better separation to always have shell scripts doing orchestration. I don't mind changing this to just use the `.R` script and call it with `Rscript`. I'll do that.

jameslamb commented on .ci/test.sh:
```diff
@@ -59,6 +59,14 @@ if [[ $TASK == "lint" ]]; then
     exit 0
 fi
 
+if [[ $TASK == "rlint" ]]; then
```
ah ok, I thought maybe it would make sense to separate them since installing R packages can kind of take a long time. I'll combine them.

jameslamb commented on .ci/test.sh:
```diff
@@ -59,6 +59,14 @@ if [[ $TASK == "lint" ]]; then
 fi
 
+if [[ $TASK == "rlint" ]]; then
+    conda install -c r \
```
oh really? I did not know that

jameslamb commented:
> @jameslamb Thanks for your PR! Please address some comments related to setup process below.

thanks for the review! Addressed the comments + rebased to `master` to get the fix from #2440

StrikerRUS commented on .ci/lint_r_code.R:
```diff
@@ -0,0 +1,64 @@
+
+results <- c()
+
+for (r_file in FILES_TO_LINT){
```
Space before `{`?

StrikerRUS commented on .ci/lint_r_code.R:
```diff
@@ -0,0 +1,64 @@
+
+issues_found <- length(results)
+
+if (issues_found > 0){
```
Space before `{`?

StrikerRUS commented on .ci/test.sh:
```diff
@@ -51,11 +51,20 @@ if [[ $TRAVIS == "true" ]] && [[ $TASK == "check-docs" ]]; then
+    conda install -q -y -n $CONDA_ENV \
+        pycodestyle \
+        pydocstyle \
+        r-argparse \
```
Is it light-weighted in terms of dependencies package? I mean, can avoid installing this library for the linting check with the aim to speed them up?

jameslamb replied:
You know, since I'm only passing a single argument I can probably just use `commandArgs()` in the script and avoid this altogether. Let me try that. One less dependency to worry about, you are right.

StrikerRUS commented on .ci/test.sh:
```diff
@@ -51,11 +51,20 @@ if [[ $TRAVIS == "true" ]] && [[ $TASK == "check-docs" ]]; then
+        r-argparse \
+        r-lintr
     pip install --user cpplint
+    echo "linting Python code"
```
Capitalize messages (all three in this file)?

jameslamb replied:
> Also, I guess, that `.Rd` files should be re-generated.

Just regenerated the files. Note that to get past the r-lib/roxygen2#822, I had to install the latest `roxygen2`. Steps to build the docs on Mac (just documenting in case anyone is curious).

```
export CXX=/usr/local/bin/g++-8 CC=/usr/local/bin/gcc-8
Rscript build_r.R

# install roxygen2 from github
Rscript -e "devtools::install_github('r-lib/roxygen2')"
cd R-package
Rscript -e "roxygen2::roxygenise(load = 'installed')"
```

StrikerRUS commented on .travis.yml:
```diff
@@ -18,6 +18,7 @@ env:
     - TASK=bdist
     - TASK=if-else
     - TASK=lint
+    - TASK=rlint
```
Remove this line.

StrikerRUS commented on .ci/lint_r_code.R:
```diff
@@ -0,0 +1,64 @@
+    , include.dirs = FALSE
+)
+
+LINTERS_TO_USE <- list(
```
Can we use more linters? For instance, `camel_case_linter`, `spaces_left_parentheses_linter`, `T_and_F_symbol_linter`, etc.

https://github.com/jimhester/lintr#available-linters

jameslamb replied:
ah ok now I see why you were asking about spaces and parentheses. I don't want to introduce `camel_case_linter` but I think these make sense to add:

* `T_and_F_symbol_linter`
* `spaces_left_parentheses_linter`
* `paren_brace_linter` (to your questions about changing `(){` to `() {`)
* `spaces_inside_linter`
* `semicolon_terminator_linter`
* `equals_na_linter`
* `function_left_parentheses_linter`

I know Jim Hester has warned in the past that some of these can make linting take a long time (since, for example, `function_left_parentheses_linter` requires actually interpreting the source code). Let me try adding these and see how it goes! Thanks for the nudge.

StrikerRUS replied:
Ah, nice! I think you should reach the agreement about the list of linters with @Laurae2.

jameslamb replied:
ok @Laurae2 here is my proposed list, could you take a look at this list and https://github.com/jimhester/lintr#available-linters and let me know if you agree / disagree? 

Once you and I agree, I'll make whatever remaining changes are needed to get the code into compliance with the linters we pick. I don't want to touch any more lines yet until we agree on the list of linters.

When I call a linter "expensive" below, I mean "requires interpretation of the code and not (in my opinion) worth that cost". These can be expensive because R is a dynamically-typed, interpreted language where people can do some very..."creative"... things that are not possible in stricter languages. Linting with `lintr` can become quite costly to run especially for the more complex linters like `extraction_operator_linter`.

**Use**

* `closed_curly_linter`
* `equals_na_linter`
* `function_left_parentheses_linter`
* `implicit_integer_linter`
* `infix_spaces_linter`
* `line_length_linter` (set to 120 just to avoid super-long lines)
* `no_tab_linter`
* `open_curly_linter`
* `paren_brace_linter`
* `semicolon_terminator_linter`
* `seq_linter`
* `spaces_inside_linter`
* `spaces_left_parentheses_linter`
* `todo_comment_linter`
* `trailing_blank_lines_linter`
* `trailing_whitespace_linter`
* `T_and_F_symbol_linter`

**Don't Use (reason why)**

* `object_usage_linter` (redundant with R CMD CHECK)
* `absolute_path_linter` (expensive and error-prone. superfluous with redundant with CRAN checks + CI we'll run on multiple OSes)
* `nonportable_path_linter` (expensive and error-prone. superfluous with redundant with CRAN checks + CI we'll run on multiple OSes)
* `pipe_continuation_linter` (irrelevant, factored out pipes in #2334 
* `assignment_linter` (expensive)
* `camel_case_linter` (expensive, not worth the effort for a project of this maturity, inconsistent with our current style)
* `commas_linter` (counters our goal of comma-first)
* `commented_code_linter` (expensive)
* `cyclocomp_linter` (expensive)
* `extraction_operator_linter` (expensive and error-prone)
* `object_length_linter` (expensive)
* `object_name_linter`: (expensive, not worth the effort for a project of this maturity)
* `single_quotes_linter` (expensive)
* `undesirable_function_linter` (unnecessary)
* `undesirable_operator_linter` (expensive, also we use `<<-` in some valid but non-standard ways)
* `unneeded_concatenation_linter` (expensive)

Laurae2 replied:
Even though `object_usage_linter` is redundant with R CMD CHECK, perhaps we should keep using it?

About `line_length_linter`, should we keep it at 120 or the standard "default" 80 character maximum per line, as it is the R "standard"?

jameslamb replied:
I think we should not use `object_usage_linter`. In the past on other projects, I've found that it is error-prone and relies on the naming of scripts. So for example functions in R are lazily evaluated and `R CMD CHECK` will not complain if you define `.do_somemthing()` in `a.R` and call it in `b.R`, but `object_usage_linter` will complain and say "undefined global variable '.do_something'".

For `line_length_linter`, in my opinion 80 is overly strict and can be counterproductive, e.g. leading to people using uninformative short variable names to stay under the limit. The common standard of 80 characters in a lot of communities is [a carryover from the earliest days of computing](https://softwareengineering.stackexchange.com/a/148729) that is not really necessary any more with modern IDEs and shells. In my experience, 100 or 120 is a better compromise between meaningful variable names and excessive complexity caused by complicated one-liners in code.

jameslamb replied:
ok @Laurae2 given discussion [here](https://github.com/microsoft/LightGBM/pull/2437#discussion_r329351521) about installation from source, here is my updated proposal

**Use**

Added #2477 to #2302 . So now @Laurae2 my proposed list of linters is:

* `closed_curly_linter`
* `infix_spaces_linter`
* `line_length_linter (120)`
* `no_tab_linter`
* `open_curly_linter`
* `spaces_inside_linter`
* `spaces_left_parentheses_linter`
* `trailing_blank_lines_linter`
* `trailing_whitespace_linter`

**Use once `lintr` does a new release**

*documented in #2477*

* `equals_na_linter`
* `function_left_parentheses_linter`
* `implicit_integer_linter`
* `:paren_brace_linter`
* `semicolon_terminator_linter`
* `seq_linter`
* `todo_comment_linter`
* `T_and_F_symbol_linter`

**Do Not Use**

*same as original post*

StrikerRUS replied:
Just want to say that I'm totally agree with @jameslamb about the 80-chars line.

jameslamb replied:
@Laurae2 could you take another look at this PR, particularly let me know if you agree with the list in [this comment](https://github.com/microsoft/LightGBM/pull/2437#discussion_r329358150)? If you agree, then I'll do a tiny bit more refactoring to satisfy all those linters, and this will be good to merge. Thanks!

jameslamb replied:
I just rebased this to latest `master` to catch up with changes over the last two weeks. It is ready for review.

StrikerRUS commented on .ci/test.sh:
```diff
@@ -51,11 +51,20 @@ if [[ $TRAVIS == "true" ]] && [[ $TASK == "check-docs" ]]; then
     cpplint --filter=-build/include_subdir,-build/header_guard,-whitespace/line_length --recursive ./src ./include || exit 0
+    echo "linting R code"
+    Rscript ${BUILD_DIRECTORY}/.ci/lint_r_code.R \
+        --source-dir ${BUILD_DIRECTORY}
```
Please add `|| exit -1` after the `Rscript` call

jameslamb replied:
We are guaranteed to get a nonzero exit code from the code I'm proposing, that is what this line of code does: https://github.com/jameslamb/LightGBM/blob/misc/r_code_style/.ci/lint_r_code.R#L64

`quit(status = 10)` (for example) causes an R script to `exit 10`

Do you want me to add the `|| exit -1` just as a source of documentation in `test.sh`?

StrikerRUS replied:
@jameslamb 
> Do you want me to add the || exit -1 just as a source of documentation in test.sh?

Yes, please do this to force Travis to fail when R script returns non-zero status. As the R script is called from the shell script, without that `|| exit -1` Travis will continue running the rest of the shell script.

jameslamb replied:
Ah I see. Sorry, not trying to be annoying about this.

Any reason we do not just set `set -e` at the top of `test.sh`? If we did that, any call exiting with a non-zero exit code would cause the script itself to exit with a non-zero and we wouldn't have to manually manage exit codes on all of these steps.

jameslamb replied:
If you are open to that I'd be happy to open a separate PR that adds a `set -e` and removes all the `|| exit -1` in this script.

StrikerRUS replied:
No problem!
> Any reason we do not just set set -e at the top of test.sh? 

IDK, let's try and see... Maybe we should allow something to fail.
But let's do not mix things: merge this PR with `|| exit -1` and then feel free to replace all `|| exit -1` with `set -e`.

jameslamb replied:
🤝 agreed!

jameslamb commented on R-package/R/lgb.interprete.R:
```diff
@@ -64,12 +64,21 @@ lgb.interprete <- function(model,
-                                FUN = function(x) matrix(seq_len(length(x)) - 1, ncol = num_class, byrow = TRUE))
+  tree_index_mat_list <- lapply(
+    X = leaf_index_mat_list
+    , FUN = function(x){
```
I usually do not use a space before `{` and the linters I'm proposing don't enforce that

jameslamb commented on R-package/R/lgb.interprete.R:
```diff
@@ -166,16 +178,19 @@ single.row.interprete <- function(tree_dt, num_class, tree_index_mat, leaf_index
-    tree_interpretation_dt <- Reduce(f = function(x, y) merge(x, y, by = "Feature", all = TRUE),
+    tree_interpretation_dt <- Reduce(
+      f = function(x, y) merge(x, y, by = "Feature", all = TRUE),
                                      x = tree_interpretation)
```
ah yes! Will fix

jameslamb commented on R-package/R/lgb.prepare.R:
```diff
@@ -1,10 +1,13 @@
 #'
-#' @return The cleaned dataset. It must be converted to a matrix format (\code{as.matrix}) for input in \code{lgb.Dataset}.
+#' @return The cleaned dataset. It must be converted to a matrix format (\code{as.matrix})
+#' for input in \code{lgb.Dataset}.
```
yep good catch!

jameslamb commented on R-package/R/lgb.prepare_rules2.R:
```diff
@@ -1,11 +1,18 @@
 #'
-#' @return A list with the cleaned dataset (\code{data}) and the rules (\code{rules}). The data must be converted to a matrix format (\code{as.matrix}) for input in \code{lgb.Dataset}.
+#' @return A list with the cleaned dataset (\code{data}) and the rules (\code{rules}).
+#'        The data must be converted to a matrix format (\code{as.matrix}) for input in
```
I was so close 😆

jameslamb commented:
> Also, I guess, that `.Rd` files should be re-generated.

Just regenerated the files. Note that to get past the r-lib/roxygen2#822, I had to install the latest `roxygen2`. Steps to build the docs on Mac (just documenting in case anyone is curious).

```
export CXX=/usr/local/bin/g++-8 CC=/usr/local/bin/gcc-8
Rscript build_r.R

# install roxygen2 from github
Rscript -e "devtools::install_github('r-lib/roxygen2')"
cd R-package
Rscript -e "roxygen2::roxygenise(load = 'installed')"
```

jameslamb commented on R-package/R/lgb.cv.R:
```diff
@@ -215,7 +229,10 @@ lgb.cv <- function(params = list(),
-      callbacks <- add.cb(callbacks, cb.early.stop(params[[which(names(params) %in% early_stop)[1]]], verbose = verbose))
+      callbacks <- add.cb(
+        callbacks
+        , cb.early.stop(params[[which(names(params) %in% early_stop)[1]]], verbose = verbose)
```
done

StrikerRUS commented:
@jameslamb Something is wrong: 
> Linting R code
Error in library(lintr) : there is no package called ‘lintr’
Execution halted
Linting C++ code

https://travis-ci.org/microsoft/LightGBM/jobs/590986051#L838-L842

BTW, as you can see, despite the R script failed, Travis `lint` job is still green an didn't stopped after a failure in the R script. That is why `|| exit -1` is needed.

StrikerRUS commented on .ci/lint_r_code.R:
```diff
@@ -0,0 +1,82 @@
+    , include.dirs = FALSE
+)
+
+# Some linters from the lintr package have not made it to CRAN yet
```
How many (approximately: 1, 2, or more) linters are not available with the package version installed from CRAN?

jameslamb replied:
it's on the order of 4-6. Yeah, I kind of wanted to test how the installation from source plays with `conda` R, and the only way I knew to reliably reproduce the environment on Travis was to just directly try on Travis.

I'll take out the unreleased ones for now and create an issue (closed and attached to #2302 ) for adding the new ones once `lintr` gets updated. I'm also going to gently open a "can you please do a release" issue on `lintr`, since it has been almost a year since the last one.

StrikerRUS replied:
Thank you very much! I think it's the most appropriate way to introduce the linting and do not suffer from it!

jameslamb replied:
I have added [a comment](https://github.com/jimhester/lintr/issues/318#issuecomment-536313549) on jimhester/lintr#318 asking for a new release.

jameslamb commented:
> @jameslamb Something is wrong:
> 
> > Linting R code
> > Error in library(lintr) : there is no package called ‘lintr’
> > Execution halted
> > Linting C++ code
> 
> https://travis-ci.org/microsoft/LightGBM/jobs/590986051#L838-L842
> 
> BTW, as you can see, despite the R script failed, Travis `lint` job is still green an didn't stopped after a failure in the R script. That is why `|| exit -1` is needed.

Thanks for catching that. Now that I've added the explicit `exit` it does now seem CI is correctly failing.

![image](https://user-images.githubusercontent.com/7608904/65835222-5a122600-e2a9-11e9-9e50-d1a5a695a4ec.png)

jameslamb commented:
ping @Laurae2 for a review

(I'll rebase to `master` and make any code changes to pass linting once we agree on the set of linters)

StrikerRUS commented on .ci/test.sh:
```diff
@@ -51,11 +51,23 @@ if [[ $TRAVIS == "true" ]] && [[ $TASK == "check-docs" ]]; then
+    Rscript ${BUILD_DIRECTORY}/.ci/lint_r_code.R ${BUILD_DIRECTORY} || exit -1
+    echo "Linting C++ code"
+    cpplint \
+        --filter=-build/include_subdir,-build/header_guard,-whitespace/line_length \
```
Bad rebase: `-build/c++11` is lost.

jameslamb replied:
oh no! sorry, good catch. will update it.

jameslamb replied:
This has been updated. I removed the `\` so the `cpplint` line is now completely untouched. Mind taking one more look? I can't merge with pending changes. Thanks!

jameslamb commented:
Thanks for the reviews everyone! I know it's just a style PR but I'm excited about this one. Making the R side of the CI stronger makes me happy 😀
